File: boj/BFSDFS/BOJ_17836.swift
/*
 백준 17836 공주님을 구해라!
 21.11.15
 https://github.com/skyqnaqna/algorithm_study
 */

import Foundation

let INF = Int(2e9)
let direction = [(-1, 0), (1, 0), (0, -1), (0, 1)]
let input = readLine()!.split(separator: " ").map { Int(String($0))! }
let n = input[0], m = input[1], t = input[2]

var gram = (-1, -1)

var graph = [[Int]](repeating: [Int](), count: n),
    dist = [[Int]](repeating: [Int](repeating: INF, count: m), count: n)

for i in 0 ..< n {
  graph[i] = readLine()!.split(separator: " ").map { Int(String($0))! }
  
  for j in 0 ..< m {
    if graph[i][j] == 2 {
      gram = (i, j)
    }
  }
}

func inBound(_ r: Int, _ c: Int) -> Bool {
  return 0 <= r && r < n && 0 <= c && c < m && graph[r][c] != 1
}

dist[0][0] = 0
var q = [(Int, Int, Int)](), front = 0
q.append((0, 0, 0))

while front < q.count {
  let (r, c, w) = q[front]
  front += 1
  
  for (dy, dx) in direction {
    let nr = r + dy
    let nc = c + dx
    
    if !inBound(nr, nc) { continue }
    
    if dist[nr][nc] == INF {
      dist[nr][nc] = w + 1
      q.append((nr, nc, w + 1))
    }
  }
}

let answer = min(dist[n-1][m-1], dist[gram.0][gram.1] + n - 1 - gram.0 + m - 1 - gram.1)

if answer > t || answer == INF {
  print("Fail")
} else {
  print(answer)
}

File: boj/BFSDFS/BOJ_17836.cpp
/*
 백준 17836 공주님을 구해라!
 21.07.22
 https://github.com/skyqnaqna/algorithm_study
 */

#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using dd = double;
using pii = pair<int, int>;
using pll = pair<ll, ll>;

#define INF 1e9
#define endl "\n"

int n, m, t, gr, gc, graph[100][100], dist[100][100];
bool visited[100][100];
int dy[] = {-1, 1, 0, 0};
int dx[] = {0, 0, -1, 1};

void bfs() {
    visited[0][0] = true;
    dist[0][0] = 0;
    queue<pair<int, pii> >q;
    q.push({0, {0, 0}});

    while (!q.empty()) {
        int d = q.front().first;
        int r = q.front().second.first;
        int c = q.front().second.second;
        q.pop();

        for (int i = 0; i < 4; ++i) {
            int nr = r + dy[i];
            int nc = c + dx[i];

            if (nr < 0 || nc < 0 || nr >= n || nc >= m) continue;

            if (!visited[nr][nc] && graph[nr][nc] != 1) {
                visited[nr][nc] = true;
                dist[nr][nc] = d + 1;
                q.push({d + 1, {nr, nc}});
            }
        }
    }
}

int main()
{
    ios::sync_with_stdio(0); cin.tie(0);
    freopen("input.txt", "r", stdin);

    cin >> n >> m >> t;

    int fromGram = 0;
    fill(&dist[0][0], &dist[99][100], INF);

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            cin >> graph[i][j];

            if (graph[i][j] == 2) {
                fromGram = n - i + m - j - 2;
                gr = i, gc = j;
            }
        }
    }

    bfs();

    int res = min(dist[n-1][m-1], dist[gr][gc] + fromGram);

    if (res <= t && res > 0) cout << res << endl;
    else cout << "Fail\n";

    return 0;
}
